pylearn2/linear/cudnn2d.py

```diff
@@ -0,0 +1,204 @@
+    output_axes : tuple, optional
+        The requested output axes. If not specified `bc01` will be used.
+    subsample : tuple or list, optional
+        Factor by which to subsample the output. Default (1, 1)
+    border_mode : string, optional
+        ‘valid’ or ‘full’. See scipy.signal.convolve2d
+    filters_shape : tuple of length 2 or 3, optional
+        ([filter's number,] filter's height, filter's width)
+    message : string, optional
+        TODO
+    """
+
+    def __init__(self,
+                 filters,
+                 batch_size,
+                 input_space,
+                 output_axes=('b', 'c', 0, 1),
+                 subsample=(1, 1),
+                 border_mode='valid',
+                 filters_shape=None,
+                 message=''):
+
+        assert batch_size is None or batch_size > 0
+        self._input_space = input_space
+        self._output_axes = output_axes
+        self._subsample = tuple(subsample)
+        self._filters = filters
+
+        super(Cudnn2D, self).__init__(
+            filters=filters,
+            img_shape=(batch_size, input_space.num_channels,
+                       input_space.shape[0], input_space.shape[1]),
+            subsample=self._subsample,
+            border_mode=border_mode,
+            filters_shape=filters.get_value(borrow=True).shape,
+            message=message
+        )
+
+        # conv_op has to be changed
+        self._conv_op = GpuDnnConv()
+        self._desc = GpuDnnConvDesc(border_mode=border_mode,
+                                    subsample=self._subsample,
+                                    conv_mode='conv')
+
+    @functools.wraps(P2LT.get_params)
+    def get_params(self):
+        return [self._filters]
+
+    @functools.wraps(P2LT.get_weights_topo)
+    def get_weights_topo(self, borrow):
+        """
+        Parameters
+        ----------
+        borrow : TODO
+            TODO
+        """
+        return np.transpose(self._filters.get_value(borrow=borrow),
+                            (0, 2, 3, 1))
+
+    def lmul(self, x):
+        """
+        .. todo::
+
+            WRITEME properly
+
+        dot(x, A)
+
+        This method overrides the original Conv2D lmul to make it work
+        with arbitrary axis orders
+
+        Parameters
+        ----------
+        x : TODO
+            TODO
+        """
+
+        # x must be formatted as batch index, channel, topo dim 0, topo dim 1
+        # for use with conv2d, so check what the current input space format is
+        assert x.ndim == 4
+        axes = self._input_space.axes
+        assert len(axes) == 4
+
+        op_axes = ('b', 'c', 0, 1)
+
+        if tuple(axes) != op_axes:
+            x = x.dimshuffle(*[axes.index(ax) for ax in op_axes])
+
+        # The calling format has to be changed
+        img = gpu_contiguous(x)
+        kerns = gpu_contiguous(self._filters)
+        desc = self._desc(img.shape, kerns.shape)
+        rval = self._conv_op(img, kerns, desc)
```

Comment:
This is not working as it is an implicit call to https://github.com/Theano/Theano/blob/fb4d5b4d50be1cda5ee4c49e9d4b88f0f2f817f3/theano/sandbox/cuda/dnn.py#L419 which needs another parameter (`output`). Can you suggest me a proper way to set `output`?